### DISCUSSION TITLE
Creating Stateful Widget

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#11003] [General] [answered]
Hi, 
I wish to create a StatefulWidget that will be able to store and restore widget data upon reload. I'm able to locate the `StatefulWidget` Interface and have implemented. What are the other steps that needs to be followed to persist data across reloads ? Any simpler example in theia would really be helpful.

Thanks in advance

custom-widget.tsx
```
@injectable()
export class CustomWidget extends ReactWidget implements StatefulWidget {

    @inject(OtherService) otherService: OtherService;
    sample :string;

    constructor() {
        super()
        this.id = 'CUSTOM_WIDGET_ID'
    }

    @postConstruct()
    init() {
        this.sample = this.otherService.getValue();
    }
    protected render() {
        return (<div className='title'>
            {this.sample}
        </div>
        )
    }

    storeState(): object {
        return { sample: this.sample }
    }
    restoreState(oldState: object): void {
        this.sample = oldState['sample'];
    }
}

```
in custom-frontend.tsx
`
    bind(ProjectTitle).toSelf().inSingletonScope();
`

**Comment by msujew** [comment, marked as answer]
Hi @maheshrajrp,

usually, just implementing the `StatefulWidget` interface should be enough to persist widget data across sessions. So, in that regard your implementation looks to be correct. What issues are you facing exactly?

**Reply by maheshrajrp** [reply]
This example seems to working fine. There seems to be issue with my implementation, will debug. 
Thanks and Sorry for the late response.
